### PR TITLE
Change hip filename extension to .hip

### PR DESCRIPTION
**Pull request by bddppq**
@xw285cornell 

- To make hip files to have unique filename extension we change hip files from _hip.cc to .hip (it's the only blessing option other than .cu in hipcc https://github.com/ROCm-Developer-Tools/HIP/blob/3d51a1fb0105e2f2312d2523c20e0034339f6ada/bin/hipcc#L552).
- Change to use host compiler to compile .cc|.cpp files. Previously we use hcc to compile them which is unnecessary
- Change the hipify script to not replace "gpu" with "hip" in the filename of the generated hipified files. Previously we do this because hcc has a bug when linking files that have same filename. We have now changed to use host linker to do linking so this is unnecessary anymore.

### CODE REVIEW

**Comment by petrex** [comment]
@@

File: tools/amd_build/pyHIPIFY/hipify_python.py [hunk 1]
```diff
@@ -744,13 +742,10 @@ def get_hip_file_path(filepath, hipify_caffe2, hip_suffix):
     #
     #   - If the file name contains "CUDA", replace it with "HIP", AND
     #
-    #   - If the file name contains "gpu", replace it with "hip".
-    #
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/hipify_python.py]: Why not this substitution anymore? It doesn't seem to have anything to do with the file extension anymore. (Are we officially asserting that HIP works correctly with filenames that are the same but in different directories? :)

File: tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py [hunk 2]
```diff
@@ -2216,19 +2216,19 @@
...
-    ("spatial_batch_norm_op_gpu_impl" , ("hip/spatial_batch_norm_op_hip_impl", API_CAFFE2)),
-    ("recurrent_network_executor_gpu" , ("hip/recurrent_network_executor_hip", API_CAFFE2)),
-    ("max_pool_with_index_gpu", ("hip/max_pool_with_index_hip", API_CAFFE2)),
-    ("THCCachingAllocator_gpu", ("hip/THCCachingAllocator_hip", API_CAFFE2)),
-    ("top_k_heap_selection", ("hip/top_k_heap_selection_hip", API_CAFFE2)),
-    ("top_k_radix_selection", ("hip/top_k_radix_selection_hip", API_CAFFE2)),
-    ("GpuDefs", ("hip/GpuDefs_hip", API_CAFFE2)),
-    ("GpuScanUtils", ("hip/GpuScanUtils_hip", API_CAFFE2)),
-    ("GpuBitonicSort", ("hip/GpuBitonicSort_hip", API_CAFFE2)),
+    ("/context_gpu" , ("/hip/context_gpu", API_CAFFE2)),
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/cuda_to_hip_mappings.py]: "Sigh, word boundaries"

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -374,14 +374,14 @@ if(USE_ROCM)
   # Call again since Caffe2_HIP_INCLUDES is extended with ATen include dirs.
   hip_include_directories(${Caffe2_HIP_INCLUDES})
 
-  filter_list(__caffe2_hip_srcs_cpp Caffe2_HIP_SRCS "\\.(cc|cpp|cu)$")
+  filter_list(__caffe2_hip_srcs_cpp Caffe2_HIP_SRCS "\\.(cu|hip)$")
```
Comment (ezyang) [inline caffe2/CMakeLists.txt]: If we have changed the hip filename extension, we shouldn't match for cu anymore, right?

File: caffe2/core/common_gpu.cc [hunk 4]
```diff
@@ -194,7 +194,7 @@ void DeviceQuery(const int device) {
      << std::endl;
   ss << "Total registers per block:     " << prop.regsPerBlock << std::endl;
   ss << "Warp size:                     " << prop.warpSize << std::endl;
-#ifndef __HIPCC__
+#ifndef __HIP_PLATFORM_HCC__
```
[the remark this answers is not in the document]
Comment (ezyang) [reply caffe2/core/common_gpu.cc]: Oh, is this because we're no longer compiling with hcc?

File: tools/amd_build/pyHIPIFY/hipify_python.py (hunk 1 above)
Comment (ezyang) [reply tools/amd_build/pyHIPIFY/hipify_python.py]: I'm actually a fan of filenames being unique, regardless of directory structure. (It's not a big deal, but there are small effects; for example, the Caffe2 enforce failure message only prints the basename of the filename, so if you disambiguated based on directory it's now ambiguous; another example, you can `find -name` with 100% reliability)

File: caffe2/core/common_gpu.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply caffe2/core/common_gpu.cc]: I am a little confused: who is responsible for setting this macro? The documentation suggests that the HIP headers are responsible for defining then. Then it is *definitely* inappropriate to use this macro in a header (because the semantics of the header changes depending on if you included the header before or after the HIP headers), and for good measure, we should probably avoid using it in non-headers as well.

Basically, if you want to add some block of code that is toggled by HIPIFY, it should be some sort of constant that the HIPIFY script replaces with the appropriate new constant; it shouldn't be conditioned on something from a HIP header.

File: caffe2/CMakeLists.txt (hunk 3 above)
Comment (ezyang) [reply caffe2/CMakeLists.txt]: Deserves a comment

File: caffe2/core/common_gpu.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bddppq) [reply caffe2/core/common_gpu.cc]: we have set this macro in cmake/Dependencies.cmake

File: tools/amd_build/pyHIPIFY/hipify_python.py (hunk 1 above)
Comment (bddppq) [reply tools/amd_build/pyHIPIFY/hipify_python.py]: making the filename has benefit (especially for headers) that potentially we can prepend the output path to include paths in the compilation commands and then we don't need to rewrites "#include" in the source files.

File: caffe2/core/common_gpu.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply caffe2/core/common_gpu.cc]: It's still inappropriate to put `#ifdef` on this macro in headers. Then the header will preprocess differently depending on which compiler invocation calls it (if you build a cc file with or without the flag.)
